src/main/java/com/google/api/codegen/gapic/GapicProviderFactory.java

```diff
@@ -0,0 +1,202 @@
+import com.google.api.codegen.csharp.CSharpSnippetSetRunner;
+import com.google.api.codegen.go.GoGapicContext;
+import com.google.api.codegen.go.GoSnippetSetRunner;
+import com.google.api.codegen.java.JavaGapicContext;
+import com.google.api.codegen.java.JavaIterableSnippetSetRunner;
+import com.google.api.codegen.java.JavaSnippetSetRunner;
+import com.google.api.codegen.nodejs.NodeJSGapicContext;
+import com.google.api.codegen.nodejs.NodeJSSnippetSetRunner;
+import com.google.api.codegen.php.PhpGapicContext;
+import com.google.api.codegen.php.PhpSnippetSetRunner;
+import com.google.api.codegen.py.PythonGapicContext;
+import com.google.api.codegen.py.PythonInterfaceInitializer;
+import com.google.api.codegen.py.PythonProtoFileInitializer;
+import com.google.api.codegen.py.PythonSnippetSetRunner;
+import com.google.api.codegen.ruby.RubyGapicContext;
+import com.google.api.codegen.ruby.RubySnippetSetRunner;
+import com.google.api.tools.framework.model.Interface;
+import com.google.api.tools.framework.model.Model;
+import com.google.api.tools.framework.model.ProtoFile;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.apache.commons.lang3.NotImplementedException;
+
+/**
+ * GapicProviderFactory creates GapicProvider instances based on an id.
+ */
+public class GapicProviderFactory {
+
+  /**
+   * Create the GapicProviders based on the given id
+   */
+  public static List<GapicProvider<? extends Object>> defaultCreate(Model model,
+      ApiConfig apiConfig, String id) {
+
+    // Please keep the following IDs in alphabetical order
+
+    if (id.equals("client_config")) {
+      GapicProvider<? extends Object> provider = CommonGapicProvider.<Interface>newBuilder()
+          .setModel(model)
+          .setView(new InterfaceView())
+          .setContext(new ClientConfigGapicContext(model, apiConfig))
+          .setSnippetSetRunner(new ClientConfigSnippetSetRunner<Interface>())
+          .setSnippetFileNames(Arrays.asList("json.snip"))
+          .build();
+      return Arrays.<GapicProvider<? extends Object>>asList(provider);
+
+    } else if (id.equals("csharp")) {
+      GapicProvider<? extends Object> provider = CommonGapicProvider.<Interface>newBuilder()
+          .setModel(model)
+          .setView(new InterfaceView())
+          .setContext(new CSharpGapicContext(model, apiConfig))
+          .setSnippetSetRunner(new CSharpSnippetSetRunner<Interface>())
+          .setSnippetFileNames(Arrays.asList("wrapper.snip"))
+          .build();
+      return Arrays.<GapicProvider<? extends Object>>asList(provider);
+
+    } else if (id.equals("go")) {
+      GapicProvider<? extends Object> provider = CommonGapicProvider.<Interface>newBuilder()
+          .setModel(model)
+          .setView(new InterfaceView())
+          .setContext(new GoGapicContext(model, apiConfig))
+          .setSnippetSetRunner(new GoSnippetSetRunner<Interface>())
+          .setSnippetFileNames(Arrays.asList("main.snip", "example.snip", "doc.snip"))
```

Comment:
It's arguable whether or not these are magic numbers. Maybe a better system would be to just scan the appropriate source directory for all .snip files. Just a thought for the future, rather than something to implement right now.